llama.cpp

```diff
@@ -8945,47 +8956,20 @@ struct llm_tokenizer_wpm {
         return words;
     }
 
-    std::string normalize(const std::string & text) {
-        // TODO: handle chinese characters? https://github.com/huggingface/tokenizers/blob/ef5f50605ddf9f8caef1598c0e4853862b9707a7/tokenizers/src/normalizers/bert.rs#L98
-        std::string text2 = strip_accents(text);
-        for (size_t i = 0; i < text2.size(); i += utf8_len(text2[i])) {
-            char c = text2[i];
-            if (c >= 'A' && c <= 'Z') {
-                text2[i] = c - 'A' + 'a';
-            }
+    uint32_t to_lower(uint32_t code) {
+#if defined(_WIN32)
+        if (code > 0xFFFF) {
+            return code;
         }
-        return text2;
+#endif
+        return std::tolower(wchar_t(code), std::locale("en_US.UTF-8"));
```

Comment:
@iamlemec This depends on the system having the `en_US.UTF-8` locale available, which is not always true in practice: https://github.com/nomic-ai/gpt4all/issues/2160

It seems like `.UTF-8` at least works on Windows (it's used by llama-bench), but on Unix I suppose you can only really assume that the plain C locale is available...

Reply:
Also, it's not safe to assume that lowercasing one letter at a time is the same as lowercasing the whole string - see the exception for capital sigma at the end of a word https://github.com/yhirose/cpp-unicodelib/blob/797b1f0f1592ce13afabf3576f51ef26db5e884d/UCD/SpecialCasing.txt#L211